daml-script/dump/src/main/scala/com/daml/script/dump/Dependencies.scala

```diff
@@ -0,0 +1,125 @@
+    def go(
+        todo: List[PackageId],
+        acc: Map[PackageId, (ByteString, Ast.Package)],
+    ): Future[Map[PackageId, (ByteString, Ast.Package)]] =
+      todo match {
+        case Nil => Future.successful(acc)
+        case p :: todo if acc.contains(p) => go(todo, acc)
+        case p :: todo =>
+          client.packageClient.getPackage(p).flatMap { pkgResp =>
+            val cos = damlLfCodedInputStreamFromBytes(
+              pkgResp.archivePayload.toByteArray,
+              Decode.PROTOBUF_RECURSION_LIMIT,
+            )
+            val pkgId = PackageId.assertFromString(pkgResp.hash)
+            val pkg = Decode
+              .readArchivePayloadAndVersion(pkgId, DamlLf.ArchivePayload.parser().parseFrom(cos))
+              ._1
+              ._2
+            go(todo ++ pkg.directDeps, acc + (pkgId -> ((pkgResp.archivePayload, pkg))))
+          }
+      }
+    go(references, Map.empty)
+  }
+
+  def writeDar(
+      sdkVersion: String,
+      file: Path,
+      dar: Dar[(PackageId, ByteString, Ast.Package)],
+  ): Unit = {
+    def encode(pkgId: PackageId, bs: ByteString) = {
+      DamlLf.Archive
+        .newBuilder()
+        .setHash(pkgId)
+        .setHashFunction(DamlLf.HashFunction.SHA256)
+        .setPayload(bs)
+        .build()
+    }
+
+    val out = new ZipOutputStream(new FileOutputStream(file.toFile))
+    out.putNextEntry(new ZipEntry("META-INF/MANIFEST.MF"))
+    out.write(manifest(sdkVersion, dar))
+    out.closeEntry()
+    dar.all.foreach { case (pkgId, bs, _) =>
+      out.putNextEntry(new ZipEntry(pkgId + ".dalf"))
+      out.write(encode(pkgId, bs).toByteArray)
+      out.closeEntry
+    }
+    out.close
+  }
+
+  // Given the pkg id of a main dalf and the map of all downloaded packages produce
+  // a DAR or return None for builtin packages like daml-stdlib
+  // that don’t need to be listed in data-dependencies.
+  def toDar(
+      pkgId: PackageId,
+      pkgs: Map[PackageId, (ByteString, Ast.Package)],
+  ): Option[Dar[(PackageId, ByteString, Ast.Package)]] = {
+    def deps(pkgId: PackageId): Set[PackageId] = {
+      @tailrec
+      def go(todo: List[PackageId], acc: Set[PackageId]): Set[PackageId] =
+        todo match {
+          case Nil => acc
+          case p :: todo if acc.contains(p) => go(todo, acc)
+          case p :: todo =>
+            go(todo ++ pkgs(p)._2.directDeps.toList, acc.union(pkgs(p)._2.directDeps))
+        }
+      go(List(pkgId), Set.empty) - pkgId
+    }
+    for {
+      pkg <- pkgs.get(pkgId) if pkg._2.metadata.isDefined
+      if !Seq("daml-stdlib", "daml-prim", "daml-script")
+        .map(Ref.PackageName.assertFromString(_))
+        .contains(pkg._2.metadata.get.name)
+    } yield {
+      Dar(
+        (pkgId, pkg._1, pkg._2),
+        deps(pkgId).toList.map(pkgId => (pkgId, pkgs(pkgId)._1, pkgs(pkgId)._2)),
+      )
+    }
+  }
+
+  private def manifest[A, B](sdkVersion: String, dar: Dar[(PackageId, A, B)]): Array[Byte] = {
```

Comment:
Don't we have some existing function that does this?

Reply:
Unfortunately not, we have the daml-lf encoder but it hardcodes a bunch of assumptions that don’t work here and it is only intended for debugging so I don’t actually want to depend on it here. I think ideally, we’d factor this and maybe the whole dependency stuff in a separate library and make the encoder depend on it.